src/preparation_graph.rs

```diff
@@ -145,13 +145,12 @@ impl PreparationGraph {
     }
 
     fn assert_valid_node_id(&self, node: NodeId) {
-        assert!(
-            node < self.num_nodes,
-            format!(
-                "invalid node id {}, must be in [0, {}[",
+        if node >= self.num_nodes {
+            panic!(
+                "invalid node id {}, must be in [0, {})",
```

Comment:
At first I thought the second `[` was a typo, but I guess you meant the open interval. I learned `[]` as closed and `()` as open in school -- not sure if this is standard or not. I can revert back to `[` if you prefer.

Reply:
Ok, not sure about North America, but over here we learn either `[a,b[` or `[a,b)` as (half) open interval, it is equivalent. Anyway, either way is fine for me.